inference-engine/tests/functional/plugin/shared/include/behavior/infer_request.hpp

```diff
@@ -623,4 +625,50 @@ TEST_P(InferRequestTests, returnDeviceBusyOnGetPerformanceCountAfterAsyncInfer)
         std::cout << "Exception" << e.what() << std::endl;
     }
 }
-}  // namespace BehaviorTestsDefinitions
\ No newline at end of file
+
+class InferRequestTestsResultNotReady : public InferRequestTests {
+};
+
+TEST_P(InferRequestTestsResultNotReady, ReturnResultNotReadyFromWaitInAsyncModeForTooSmallTimeout) {
+    // Skip test according to plugin specific disabledTestPatterns() (if any)
+    SKIP_IF_CURRENT_TEST_IS_DISABLED()
+    // Create CNNNetwork from ngraph::Function
+    // return function which computes around 20ms on GNA SW
+    function = LayerTestsDefinitions::Basic_LSTM_S::GetNetwork(3000, 380);
+    InferenceEngine::CNNNetwork cnnNet(function);
+    // Load CNNNetwork to target plugins
+    auto execNet = ie->LoadNetwork(cnnNet, targetDevice, configuration);
+    // Create InferRequest
+    InferenceEngine::InferRequest req;
+    ASSERT_NO_THROW(req = execNet.CreateInferRequest());
+    InferenceEngine::ResponseDesc response;
+    InferenceEngine::StatusCode sts = InferenceEngine::StatusCode::OK;
+    std::chrono::system_clock::time_point callbackTimeStamp;
+    volatile bool callbackTimeStampSet = false;
+    std::condition_variable callbackCompleted;
+    std::mutex m;
+    // add a callback to the request and capture the timestamp
+    req.SetCompletionCallback([&]() {
+        std::unique_lock<std::mutex> lock(m);
+        callbackTimeStamp = std::chrono::system_clock::now();
+        callbackTimeStampSet = true;
+        lock.unlock();
+        callbackCompleted.notify_one();
+        });
+    req.StartAsync();
+    ASSERT_NO_THROW(sts = req.Wait(InferenceEngine::IInferRequest::WaitMode::STATUS_ONLY));
+    // get timestamp taken AFTER return from the Wait(STATUS_ONLY)
+    const auto afterWaitTimeStamp = std::chrono::system_clock::now();
+    {
+        // wait for callback to complete
+        std::unique_lock<std::mutex> lock(m);
+        callbackCompleted.wait(lock, [&] {return callbackTimeStampSet; });
```

Comment:
I guess this can be replaced with 
while(!callBackTimeStepSet) {
}
 and you should not need any mutex/cond_var then.

Other than that the test seems to be valid

Reply:
Simplified even further with promise/future.